File: algorithms/algorithms-cplusplus/leetcode/Question_0140_Word_Break_II.cc
#include <string.h>
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;

class Solution {
 public:
  vector<string> wordBreak(string s, vector<string>& wordDict) {
    unordered_map<string, vector<string>> m;
    return help(s, wordDict, m);
  }

  vector<string> help(string s, vector<string>& wordDict, unordered_map<string, vector<string>>& m) {
    if (m.count(s)) {
      return m[s];
    }
    vector<string> res;
    if (s.empty()) {
      res.push_back("");
      return res;
    }
    for (string& word : wordDict) {
      if (word.size() > s.size()) {
        continue;
      }
      if (strncmp(s.c_str(), word.c_str(), word.size()) == 0) {
        vector<string> sub = help(s.substr(word.size()), wordDict, m);
        for (string& w : sub) {
          if (w.empty()) {
            res.push_back(word);
          } else {
            res.push_back(word + " " + w);
          }
        }
      }
    }
    m[s] = res;
    return res;
  }
};

File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_0140_Word_Break_II.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class Question_0140_Word_Break_II {
    public List<String> wordBreak(String s, List<String> wordDict) {
        return help(s, wordDict, new HashMap<String, List<String>>());
    }

    private List<String> help(String s, List<String> wordDict, HashMap<String, List<String>> map) {
        if (map.containsKey(s)) {
          return map.get(s);
        }
        List<String> res = new ArrayList<String>();
        if (s.length() == 0) {
            res.add("");
            return res;
        }
        for (String word : wordDict) {
            if (s.startsWith(word)) {
                List<String> sublist = help(s.substring(word.length()), wordDict, map);
                for (String sub : sublist) {
                    if (sub.isEmpty()) {
                        res.add(word);
                    } else {
                        res.add(word + " " + sub);
                    }
                }
            }
        }
        map.put(s, res);
        return res;
    }
}
